test/model/message_list_test.dart

```diff
@@ -162,5 +162,107 @@ void main() async {
       await checkRenderingOnly(legacy: true);
     });
+
+    group('ReactionEvent handling', () {
+      ReactionEvent mkEvent(Reaction reaction, ReactionOp op, int messageId) {
+        return ReactionEvent(
+          id: 1,
+          op: op,
+          emojiName: reaction.emojiName,
+          emojiCode: reaction.emojiCode,
+          reactionType: reaction.reactionType,
+          userId: reaction.userId,
+          messageId: messageId,
+        );
+      }
+
+      test('add reaction', () async {
+        final originalMessage = eg.streamMessage(id: 243, stream: stream, reactions: []);
+        final messageList = await messageListViewWithMessages([originalMessage], stream, narrow);
+
+        final message = messageList.messages.single;
+        check(message).reactions.not(it()..jsonEquals([eg.unicodeEmojiReaction]));
+
+        bool listenersNotified = false;
+        messageList.addListener(() { listenersNotified = true; });
+
+        messageList.maybeUpdateMessageReactions(
+          mkEvent(eg.unicodeEmojiReaction, ReactionOp.add, originalMessage.id));
+
+        check(listenersNotified).isTrue();
+        check(messageList.messages.single)
+          ..identicalTo(message)
+          ..reactions.jsonEquals([eg.unicodeEmojiReaction]);
+      });
+
+      test('add reaction; message is not in list', () async {
+        final someMessage = eg.streamMessage(id: 1, reactions: []);
+        final messageList = await messageListViewWithMessages([someMessage], stream, narrow);
+        check(messageList.messages.single).reactions.jsonEquals([]);
+
+        bool listenersNotified = false;
+        messageList.addListener(() { listenersNotified = true; });
+
+        messageList.maybeUpdateMessageReactions(
+          mkEvent(eg.unicodeEmojiReaction, ReactionOp.add, 1000));
+
+        check(listenersNotified).isFalse();
+        check(messageList.messages.single).reactions.jsonEquals([]);
+      });
+
+      test('remove reaction', () async {
+        final eventReaction = Reaction(reactionType: ReactionType.unicodeEmoji,
+          emojiName: 'wave',                  emojiCode: '1f44b', userId: 1);
+
+        // Same emoji, different user. Not to be removed.
+        final reaction2 = Reaction.fromJson(eventReaction.toJson()..update('user_id', (_) => 2));
+
+        // Same user, different emoji. Not to be removed.
+        final reaction3 = Reaction.fromJson(eventReaction.toJson()
+          ..update('emoji_code', (_) => '1f6e0')
+          ..update('emoji_name', (_) => 'working_on_it'));
+
+        // Same user, same emojiCode, different emojiName. To be removed: servers
+        // key on user, message, reaction type, and emoji code, but not emoji name,
+        // so we follow that:
+        //   https://github.com/zulip/zulip-flutter/pull/256#discussion_r1284865099
+        final reaction4 = Reaction.fromJson(eventReaction.toJson()
+          ..update('emoji_name', (_) => 'tools'));
```

Comment:
I think this aspect doesn't affect the substance of the test, but for realism (which can be helpful for a reader): `tools` should correspond to the same emoji code as `working_on_it`, rather than the same as `wave`.

Reply:
We want the `emojiCode` to match the one in the event, but we want `emojiName` not to match, right. Since this is the reaction in the event—

```dart
        final eventReaction = Reaction(reactionType: ReactionType.unicodeEmoji,
          emojiName: 'wave',                  emojiCode: '1f44b', userId: 1);
```

—how about this for `reaction4`:

```dart
        final reaction4 = Reaction.fromJson(eventReaction.toJson()
          ..update('emoji_name', (_) => 'hello'));
```

? 🙂